### DISCUSSION TITLE
Capacity helper

### DISCUSSION THREAD

**Discussion by yoe** [home-assistant/architecture#1108] [General]
I was pointed here from the #developers discussion on discord.

I've got prepaid electricity at home, and a Shelly EM power meter which measures how much electricity is used. However, that doesn't tell me how much prepaid electricity is left. I would like to be able to know this.

I see at https://community.home-assistant.io/t/pre-paid-electricity-counter/442952 that I'm not the only person with this issue. There are a few other, similar, threads in the forums that try to solve the same issue. Although a few options exist to make this possible, they all are either a bit convoluted (requiring calculations outside of home assistant, e.g. in Node Red or through calculations on an ESP device) or very coarsely grained (the solution I link to above only calculates the remaining kWh once daily).

I was thinking that a "capacity" helper integration could solve this issue, as well as a number of other things, in a more elegant fashion. What I propose is a capacity integration that would have the following services and states:

- A source entity, which can be anything (similar to the source entity in the threshold helper)
- A "target" state, which is the current point of the source entity where the helper will report "empty"
- A "remaining" state, which will be calculated as ((source entity value) - (target state))
- A "set_target" service, which sets the target state to the given value
- A "set_capacity" service, which sets the target state to ((source entity value) + (given value))

This could be used for all kinds of prepaid services (such as prepaid electricity) where use is measured by Home Assistant; but it could also be used for other kinds of uses, such as:

- A device that requires regular maintenance after a number of times it has been used: set the capacity helper to the number of uses before maintenance is required, and create an automation to set a TODO list item and/or a notification when maintenance is due
- A device that needs to be refilled or replaced (e.g., a gas tank) for which a usage meter is available but no capacity sensor
- When used with a time value as the source entity, it can be used as a timer: set the helper's target value to a point in the future, and call an automation when the remaining state gets to be less than or equal to zero (i.e., when the target time has been reached). This can be used to replace long-running scripts and/or automations, and avoid the issue that occurs when HA is restarted.

I'm happy to implement this myself if the design seems sound, although it might take a while as I've never written anything for home assistant before, and my Python skills are reasonably limited.

**Comment by AJediIAm** [comment]
Option 1: This might be a good use case for a blue print. 
There are several examples of a template sensor to record the capacity of a water softener based on the amount of used water, but a quick search did not result in a blueprint.

Option 2: It could be an extension or variant of the utility meter integration.
Both its functionality and use case are pretty close to the capacity. The utility meter should be able to track the percentage based on the following:
- an entity to track
- the amount between 0% and 100% 
- a service call to reset to 0%
For example: 
The energy entity is at 10.000 when the reset call is made, resetting the utility/capacity meter to 0%
The utility meter capacity is set to 500.
A few hours later, the energy entity has reached 10.300
The utility/capacity meter shows 60%.

Another advantage of having a reset service call is that the current utility meter entity can be used to track an entity based in irregular intervals.

I'm not an expert, so take my 2 cents with a grain of salt.

**Reply by yoe** [reply]
Your suggestion of using percentages and preset capacities could work for some of my proposed use cases, but not all.

For instance, for prepaid electricity where I live, one does not buy a preset amount of electricity; instead, you spend however much money you wish to spend, the amount of which is then converted (using the price per kWh) into a (fractional) number of kWh units that is then added to the current amount of electricity on the meter. This could mean that you're buying 376.34 kWh one time, and 24.53 kWh the other one, depending on how much you have budget for.

**Reply by AJediIAm** [reply]
Yes. It's always a balance between simple use and powerful capabilities. Additional parameters can be added or the reset function can be given an amount for the next interval.

The more I think about it, the more I'm leaning towards a blueprint though.
